File: src/main/java/org/snakeyaml/engine/v2/serializer/Serializer.java
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.snakeyaml.engine.v2.serializer;

import org.snakeyaml.engine.v2.api.DumpSettings;
import org.snakeyaml.engine.v2.comments.CommentLine;
import org.snakeyaml.engine.v2.common.Anchor;
import org.snakeyaml.engine.v2.emitter.Emitable;
import org.snakeyaml.engine.v2.events.AliasEvent;
import org.snakeyaml.engine.v2.events.CommentEvent;
import org.snakeyaml.engine.v2.events.DocumentEndEvent;
import org.snakeyaml.engine.v2.events.DocumentStartEvent;
import org.snakeyaml.engine.v2.events.ImplicitTuple;
import org.snakeyaml.engine.v2.events.MappingEndEvent;
import org.snakeyaml.engine.v2.events.MappingStartEvent;
import org.snakeyaml.engine.v2.events.ScalarEvent;
import org.snakeyaml.engine.v2.events.SequenceEndEvent;
import org.snakeyaml.engine.v2.events.SequenceStartEvent;
import org.snakeyaml.engine.v2.events.StreamEndEvent;
import org.snakeyaml.engine.v2.events.StreamStartEvent;
import org.snakeyaml.engine.v2.nodes.AnchorNode;
import org.snakeyaml.engine.v2.nodes.MappingNode;
import org.snakeyaml.engine.v2.nodes.Node;
import org.snakeyaml.engine.v2.nodes.NodeTuple;
import org.snakeyaml.engine.v2.nodes.NodeType;
import org.snakeyaml.engine.v2.nodes.ScalarNode;
import org.snakeyaml.engine.v2.nodes.SequenceNode;
import org.snakeyaml.engine.v2.nodes.Tag;

import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

public class Serializer {
    private final DumpSettings settings;
    private final Emitable emitable;
    private final Set<Node> serializedNodes;
    private final Map<Node, Anchor> anchors;

    public Serializer(DumpSettings settings, Emitable emitable) {
        this.settings = settings;
        this.emitable = emitable;
        this.serializedNodes = new HashSet();
        this.anchors = new HashMap();
    }

    public void serialize(Node node) {
        this.emitable.emit(new DocumentStartEvent(settings.isExplicitStart(), settings.getYamlDirective(), settings.getTagDirective()));
        anchorNode(node);
        settings.getExplicitRootTag().ifPresent(node::setTag);
        serializeNode(node);
        this.emitable.emit(new DocumentEndEvent(settings.isExplicitEnd()));
        this.serializedNodes.clear();
        this.anchors.clear();
    }

    public void open() {
        this.emitable.emit(new StreamStartEvent());
    }

    public void close() {
        this.emitable.emit(new StreamEndEvent());
        // clean up the resources
        this.anchors.clear();
        this.serializedNodes.clear();
    }

    private void anchorNode(Node node) {
        final Node realNode;
        if (node.getNodeType() == NodeType.ANCHOR) {
            realNode = ((AnchorNode) node).getRealNode();
        } else {
            realNode = node;
        }
        if (this.anchors.containsKey(realNode)) {
            //it looks weird, anchors does contain the key node but we call computeIfAbsent()
            // this is because the value is null (HashMap permits values to be null)
            this.anchors.computeIfAbsent(realNode, a -> settings.getAnchorGenerator().nextAnchor(realNode));
        } else {
            this.anchors.put(realNode, realNode.getAnchor().isPresent() ? settings.getAnchorGenerator().nextAnchor(realNode) : null);
            switch (realNode.getNodeType()) {
                case SEQUENCE:
                    SequenceNode seqNode = (SequenceNode) realNode;
                    List<Node> list = seqNode.getValue();
                    for (Node item : list) {
                        anchorNode(item);
                    }
                    break;
                case MAPPING:
                    MappingNode mappingNode = (MappingNode) realNode;
                    List<NodeTuple> map = mappingNode.getValue();
                    for (NodeTuple object : map) {
                        Node key = object.getKeyNode();
                        Node value = object.getValueNode();
                        anchorNode(key);
                        anchorNode(value);
                    }
                    break;
                default: // no further action required for non-collections
            }
        }
    }

    private void serializeNode(Node node) {
        if (node.getNodeType() == NodeType.ANCHOR) {
            node = ((AnchorNode) node).getRealNode();
        }
        Optional<Anchor> tAlias = Optional.ofNullable(this.anchors.get(node));
        if (this.serializedNodes.contains(node)) {
            this.emitable.emit(new AliasEvent(tAlias));
        } else {
            this.serializedNodes.add(node);
            switch (node.getNodeType()) {
                case SCALAR:
                    ScalarNode scalarNode = (ScalarNode) node;
                    serializeComments(node.getBlockComments());
                    Tag detectedTag = settings.getScalarResolver().resolve(scalarNode.getValue(), true);
                    Tag defaultTag = settings.getScalarResolver().resolve(scalarNode.getValue(), false);
                    ImplicitTuple tuple = new ImplicitTuple(node.getTag().equals(detectedTag), node
                            .getTag().equals(defaultTag));
                    ScalarEvent event = new ScalarEvent(tAlias, Optional.of(node.getTag().getValue()), tuple,
                            scalarNode.getValue(), scalarNode.getScalarStyle());
                    this.emitable.emit(event);
                    serializeComments(node.getInLineComments());
                    serializeComments(node.getEndComments());
                    break;
                case SEQUENCE:
                    SequenceNode seqNode = (SequenceNode) node;
                    serializeComments(node.getBlockComments());
                    boolean implicitS = node.getTag().equals(Tag.SEQ);
                    this.emitable.emit(new SequenceStartEvent(tAlias, Optional.of(node.getTag().getValue()),
                            implicitS, seqNode.getFlowStyle()));
                    List<Node> list = seqNode.getValue();
                    for (Node item : list) {
                        serializeNode(item);
                    }
                    this.emitable.emit(new SequenceEndEvent());
                    serializeComments(node.getInLineComments());
                    serializeComments(node.getEndComments());
                    break;
                default:// instance of MappingNode
                    serializeComments(node.getBlockComments());
                    boolean implicitM = node.getTag().equals(Tag.MAP);
                    MappingNode mappingNode = (MappingNode) node;
                    List<NodeTuple> map = mappingNode.getValue();
                    if (mappingNode.getTag() != Tag.COMMENT) {
                        this.emitable.emit(new MappingStartEvent(tAlias, Optional.of(mappingNode.getTag().getValue()),
                                implicitM, mappingNode.getFlowStyle(), Optional.empty(), Optional.empty()));
                        for (NodeTuple entry : map) {
                            Node key = entry.getKeyNode();
                            Node value = entry.getValueNode();
                            serializeNode(key);
                            serializeNode(value);
                        }
                        this.emitable.emit(new MappingEndEvent());
                        serializeComments(node.getInLineComments());
                        serializeComments(node.getEndComments());
                    }
            }
        }
    }

    private void serializeComments(List<CommentLine> comments) {
        if (comments == null) {
            return;
        }
        for (CommentLine line : comments) {
            CommentEvent commentEvent = new CommentEvent(line.getCommentType(), line.getValue(), line.getStartMark(),
                    line.getEndMark());
            this.emitable.emit(commentEvent);
        }
    }
}

File: src/main/java/org/snakeyaml/engine/v2/serializer/Serializer.kt
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.snakeyaml.engine.v2.serializer

import org.snakeyaml.engine.v2.api.DumpSettings
import org.snakeyaml.engine.v2.comments.CommentLine
import org.snakeyaml.engine.v2.common.Anchor
import org.snakeyaml.engine.v2.emitter.Emitable
import org.snakeyaml.engine.v2.events.AliasEvent
import org.snakeyaml.engine.v2.events.CommentEvent
import org.snakeyaml.engine.v2.events.DocumentEndEvent
import org.snakeyaml.engine.v2.events.DocumentStartEvent
import org.snakeyaml.engine.v2.events.ImplicitTuple
import org.snakeyaml.engine.v2.events.MappingEndEvent
import org.snakeyaml.engine.v2.events.MappingStartEvent
import org.snakeyaml.engine.v2.events.ScalarEvent
import org.snakeyaml.engine.v2.events.SequenceEndEvent
import org.snakeyaml.engine.v2.events.SequenceStartEvent
import org.snakeyaml.engine.v2.events.StreamEndEvent
import org.snakeyaml.engine.v2.events.StreamStartEvent
import org.snakeyaml.engine.v2.exceptions.Mark
import org.snakeyaml.engine.v2.nodes.AnchorNode
import org.snakeyaml.engine.v2.nodes.MappingNode
import org.snakeyaml.engine.v2.nodes.Node
import org.snakeyaml.engine.v2.nodes.NodeType
import org.snakeyaml.engine.v2.nodes.ScalarNode
import org.snakeyaml.engine.v2.nodes.SequenceNode
import org.snakeyaml.engine.v2.nodes.Tag
import java.util.Optional

/**
 * Transform a [Node] Graph to [org.snakeyaml.engine.v2.events.Event] stream and allow provided [Emitable] to present the
 * [org.snakeyaml.engine.v2.events.Event]s into the output stream
 *
 * @param settings - dump configuration
 * @param emitable - destination for the event stream
 */
class Serializer(
    private val settings: DumpSettings,
    private val emitable: Emitable,
) {
    private val serializedNodes: MutableSet<Node> = mutableSetOf()
    private val anchors: MutableMap<Node, Anchor?> = mutableMapOf()

    /**
     * Serialize document
     *
     * @param node - the document root
     */
    fun serializeDocument(node: Node) {
        emitable.emit(
            DocumentStartEvent(
                settings.isExplicitStart,
                settings.yamlDirective,
                settings.tagDirective,
            ),
        )
        anchorNode(node)
        settings.explicitRootTag.ifPresent { tag -> node.tag = tag }
        serializeNode(node)
        emitable.emit(DocumentEndEvent(settings.isExplicitEnd))
        serializedNodes.clear()
        anchors.clear()
    }

    /** Emit [StreamStartEvent] */
    fun emitStreamStart() {
        emitable.emit(StreamStartEvent())
    }

    /** Emit [StreamEndEvent] */
    fun emitStreamEnd() {
        emitable.emit(StreamEndEvent())
    }

    private fun anchorNode(node: Node) {
        val realNode = if (node is AnchorNode) node.realNode else node

        if (anchors.containsKey(realNode)) {
            // it looks weird, anchor does contain the key node, but we call computeIfAbsent()
            // this is because the value is null (HashMap permits values to be null)
            anchors.computeIfAbsent(realNode) { settings.anchorGenerator.nextAnchor(realNode) }
        } else {
            anchors[realNode] =
                if (realNode.anchor.isPresent) {
                    settings.anchorGenerator.nextAnchor(realNode)
                } else {
                    null
                }
            when (realNode.nodeType) {
                NodeType.SEQUENCE                -> {
                    require(realNode is SequenceNode)
                    for (item in realNode.value) {
                        anchorNode(item)
                    }
                }

                NodeType.MAPPING                 -> {
                    require(realNode is MappingNode)
                    for ((key, value) in realNode.value) {
                        anchorNode(key)
                        anchorNode(value)
                    }
                }

                NodeType.SCALAR, NodeType.ANCHOR -> {}
            }
        }
    }

    /**
     * Recursive serialization of a [Node]
     *
     * @param node - content
     */
    private fun serializeNode(node: Node) {
        val realNode = if (node is AnchorNode) node.realNode else node

        val tAlias = Optional.ofNullable(anchors[realNode])
        if (serializedNodes.contains(realNode)) {
            emitable.emit(AliasEvent(tAlias))
        } else {
            serializedNodes.add(realNode)
            when (realNode.nodeType) {
                NodeType.SCALAR   -> {
                    require(realNode is ScalarNode)
                    serializeComments(realNode.blockComments)
                    val detectedTag = settings.schema.scalarResolver.resolve(realNode.value, true)
                    val defaultTag = settings.schema.scalarResolver.resolve(realNode.value, false)
                    val tuple = ImplicitTuple(
                        plain = realNode.tag == detectedTag,
                        nonPlain = realNode.tag == defaultTag,
                    )
                    val event = ScalarEvent(
                        anchor = tAlias,
                        tag = Optional.of(realNode.tag.value),
                        implicit = tuple,
                        value = realNode.value,
                        scalarStyle = realNode.scalarStyle,
                    )
                    emitable.emit(event)
                    serializeComments(realNode.inLineComments)
                    serializeComments(realNode.endComments)
                }

                NodeType.SEQUENCE -> {
                    require(realNode is SequenceNode)
                    serializeComments(realNode.blockComments)
                    val implicitS = realNode.tag == Tag.SEQ
                    emitable.emit(
                        SequenceStartEvent(
                            anchor = tAlias,
                            tag = Optional.of(realNode.tag.value),
                            implicit = implicitS,
                            flowStyle = realNode.flowStyle,
                        ),
                    )
                    for (item in realNode.value) {
                        serializeNode(item)
                    }
                    emitable.emit(SequenceEndEvent())
                    serializeComments(realNode.inLineComments)
                    serializeComments(realNode.endComments)
                }

                NodeType.MAPPING  -> {
                    require(realNode is MappingNode)
                    serializeComments(realNode.blockComments)
                    if (realNode.tag != Tag.COMMENT) {
                        emitable.emit(
                            MappingStartEvent(
                                anchor = tAlias,
                                tag = Optional.of(realNode.tag.value),
                                implicit = realNode.tag == Tag.MAP,
                                flowStyle = realNode.flowStyle,
                                startMark = Optional.empty<Mark>(),
                                endMark = Optional.empty<Mark>(),
                            ),
                        )
                        for ((key, value) in realNode.value) {
                            serializeNode(key)
                            serializeNode(value)
                        }
                        emitable.emit(MappingEndEvent())
                        serializeComments(realNode.inLineComments)
                        serializeComments(realNode.endComments)
                    }
                }

                else              -> {
                    error("Could not handle node type ${realNode.nodeType}")
                }
            }
        }
    }

    private fun serializeComments(comments: List<CommentLine>?) {
        if (comments == null) return

        for (line in comments) {
            val commentEvent = CommentEvent(
                commentType = line.commentType,
                value = line.value,
                startMark = line.startMark,
                endMark = line.endMark,
            )
            emitable.emit(commentEvent)
        }
    }
}
